js/modules/k6/experimental/fs/module.go

```diff
@@ -173,3 +178,158 @@ func (f *File) Stat() *goja.Promise {
 
 	return promise
 }
+
+// Read the file's content, and writes it into the provided Uint8Array.
+//
+// Resolves to either the number of bytes read during the operation
+// or EOF (null) if there was nothing more to read.
+//
+// It is possible for a read to successfully return with 0 bytes.
+// This does not indicate EOF.
+//
+//nolint:funlen
+func (f *File) Read(into goja.Value) *goja.Promise {
+	// This method performs an asynchronous read operation and modifies the provided Uint8Array in place.
+	// To ensure thread safety and avoid concurrency issues, we take special precautions when creating the promise:
+	//
+	// 1. Instead of using the standard [promises.New] method, we manually create a promise.
+	// 2. We register a callback to be executed by the VU's runtime. This ensures that the modification
+	//    of the JS runtime's `buffer` occurs on the main thread during the promise's resolution.
+	promise, resolveFunc, rejectFunc := f.vu.Runtime().NewPromise()
+	callback := f.vu.RegisterCallback()
+
+	resolve := func(result any) {
+		callback(func() error {
+			resolveFunc(result)
+			return nil
+		})
+	}
+
+	reject := func(reason any) {
+		callback(func() error {
+			rejectFunc(reason)
+			return nil
+		})
+	}
+
+	if common.IsNullish(into) {
+		reject(newFsError(TypeError, "read() failed; reason: into cannot be null or undefined"))
+		return promise
+	}
+
+	// We expect the into argument to be a `Uint8Array` instance
+	intoObj := into.ToObject(f.vu.Runtime())
+	uint8ArrayConstructor := f.vu.Runtime().Get("Uint8Array")
+	if isUint8Array := intoObj.Get("constructor").SameAs(uint8ArrayConstructor); !isUint8Array {
+		reject(newFsError(TypeError, "read() failed; reason: into argument must be a Uint8Array"))
+		return promise
+	}
+
+	// Obtain the underlying ArrayBuffer from the Uint8Array
+	ab, ok := intoObj.Get("buffer").Export().(goja.ArrayBuffer)
+	if !ok {
+		reject(newFsError(TypeError, "read() failed; reason: into argument cannot be interpreted as ArrayBuffer"))
+		return promise
+	}
+
+	// Copy the ArrayBuffer into a byte slice, so that we can pass it to the
+	// [file.Read] method without risking to modify the original ArrayBuffer, and
+	// running into concurrency issues (data race).
+	intoBytes := ab.Bytes()
+	buffer := make([]byte, len(intoBytes))
+
+	go func() {
+		n, err := f.file.Read(buffer)
+		if err == nil {
+			// Although the read operation happens as part of the goroutine, we
+			// still need to make sure that:
+			//   1. Any side effects, like modifying the `buffer`, are deferred and
+			//   executed on the main thread via the registered callback.
+			//   2. This approach ensures that while the file read operation can proceed
+			//   asynchronously, any side effects that might interfere with the JS runtime
+			//   are executed in a controlled and sequential manner on the main thread.
+			callback(func() error {
+				_ = copy(intoBytes, buffer)
+				resolveFunc(n)
+				return nil
+			})
+			return
+		}
```

Comment:
Currently, if there is a read with big enough buffer to read the whole file +1 byte - the buffer will not be updated.

```javascript
import fs from "k6/experimental/fs"
let f;
fs.open("./numbers.txt").then((s) => f = s);
export default async function() {
	let b = new Uint8Array(12); // works with 11 
	let n = await f.read(b)
	console.log(n)
	console.log(b)
}
```

With a numbers.txt

```
0123456789
```

Looking at the code and what the remaining one do I think that the easiest solution is to just
1. check `n == 0` here
2. not resolve here, but go to the err code and let it does it. (checking for nil as well)
3. Actually wrap the whole code after the `Read` in callback and use the original promise's resolveFunc and rejectFUnc directly.
4. due to the above -you no longer need the wrappers around resolve and rejectFunc as all calls are now either in a callback or on the eventloop already (before we return from the parent function)

Also obviously a test for this will be nice

Reply:
Good catch indeed!

I have addressed the issue, and added a dedicated test in the following commit: f5b47a21dc72a88a5862639c949af0e66cf6fad7

Let me know if that works for you 🙇🏻